src/lua/uri.lua

```diff
@@ -220,6 +267,57 @@ local function values(...)
     return {...}
 end
 
+local char_to_hex = function(c)
+    return string.format("%%%02X", string.byte(c))
+end
+
+-- Encodes a string into its escaped hexadecimal representation.
+local function escape(buf, opts)
+    if type(buf) ~= "string" then
+        error("Usage: uri.escape(string, [opts])")
+    end
+    local options = opts or RFC3986
+    options.unreserved = options.unreserved or RFC3986.unreserved
+    options.plus = options.plus or RFC3986.plus
+
+    local pattern = unreserved(options.unreserved)
+    local unreserved = {}
+    for i = 0, 255 do
+        local ch = string.char(i)
+        unreserved[ch] = ch:match(pattern)
+    end
+
+    return (buf:gsub("(.)", function(ch)
+        if options.plus and ch == " " then
+            return "+"
+        elseif unreserved[ch] then
+            return ch
+        else
+            return char_to_hex(ch)
+        end
+    end))
+end
+
+local hex_to_char = function(x)
+    return string.char(tonumber(x, 16))
+end
+
+-- Decodes an escaped hexadecimal string into its binary representation.
+local function unescape(buf, opts)
+    if type(buf) ~= "string" then
+        error("Usage: uri.unescape(string, [opts])")
+    end
+    opts = opts or RFC3986
+    opts.plus = opts.plus or RFC3986.plus
+
+    local str = buf
+    if opts.plus == true then
+        str = str:gsub("+", " ")
+    end
+    str = str:gsub("%%(%x%x)", hex_to_char)
```

Comment:
AFAIR, the standard says that an implementation should accept escaped sequences as in the lower case as well as in the upper case. Given implementation conforms this rule. I would like to hold this property in a test case.